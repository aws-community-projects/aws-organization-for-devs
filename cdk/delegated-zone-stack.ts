import { CfnOutput, Fn, RemovalPolicy, SecretValue, Stack, StackProps } from 'aws-cdk-lib';
import { Certificate, CertificateValidation } from 'aws-cdk-lib/aws-certificatemanager';
import { Role } from 'aws-cdk-lib/aws-iam';
import { CrossAccountZoneDelegationRecord, PublicHostedZone } from 'aws-cdk-lib/aws-route53';
import { Secret } from 'aws-cdk-lib/aws-secretsmanager';
import { Construct } from 'constructs';

interface DelegatedZoneStackProps extends StackProps {
  name: string;
  rootAccountId: string;
}

export class DelegatedZoneStack extends Stack {
  constructor(scope: Construct, id: string, props: DelegatedZoneStackProps) {
    super(scope, id, props);

    const { name, rootAccountId } = props;

    const domainName = `${name}.awscommunitybuilders.org`;

    const delegatedZone = new PublicHostedZone(this, 'delegatedZone', {
      zoneName: domainName,
    });

    const delegationRoleArn = Stack.of(this).formatArn({
      account: rootAccountId,
      region: '',
      resource: 'role',
      resourceName: 'HostedZoneDelegationRole',
      service: 'iam',
    });
    const delegationRole = Role.fromRoleArn(this, 'DelegationRole', delegationRoleArn);

    new CrossAccountZoneDelegationRecord(this, 'DelegationRecord', {
      delegationRole,
      delegatedZone,
      parentHostedZoneName: 'awscommunitybuilders.org',
    });

    const certificate = new Certificate(this, 'DelegatedZoneCert', {
      domainName: `*.${domainName}`,
      validation: CertificateValidation.fromDns(delegatedZone),
    });

    new Secret(this, 'AccountCerteArn', {
      removalPolicy: RemovalPolicy.DESTROY,
      replicaRegions: [{ region: 'eu-west-1' }],
      secretName: '/account/certificateArn',
      secretStringValue: SecretValue.unsafePlainText(certificate.certificateArn),
    });

    new CfnOutput(this, 'nameservers', {
      description: 'nameservers',
      value: Fn.join(', ', delegatedZone.hostedZoneNameServers!),
    });

    new CfnOutput(this, 'certArn', {
      description: 'certificate arn',
      value: certificate.certificateArn,
    });
  }
}
