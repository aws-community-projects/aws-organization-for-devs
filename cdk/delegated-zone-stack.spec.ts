import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { describe, expect, test } from 'vitest';

import { DelegatedZoneStack } from './delegated-zone-stack';

describe('Entire Stack', () => {
  test('match a snapshot', () => {
    const app = new App();
    const stack = new DelegatedZoneStack(app, 'test-stack');
    const template = Template.fromStack(stack);

    expect(template).toMatchSnapshot();
  });
});
