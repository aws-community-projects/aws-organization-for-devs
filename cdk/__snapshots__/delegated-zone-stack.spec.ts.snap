// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Entire Stack > match a snapshot 1`] = `
{
  "Outputs": {
    "certArn": {
      "Description": "certificate arn",
      "Value": {
        "Ref": "DelegatedZoneCert71201BFD",
      },
    },
    "nameservers": {
      "Description": "nameservers",
      "Value": {
        "Fn::Join": [
          ", ",
          {
            "Fn::GetAtt": [
              "delegatedZone93B2A299",
              "NameServers",
            ],
          },
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AccountCerteArn7BFDEF53": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Name": "/account/certificateArn",
        "ReplicaRegions": [
          {
            "Region": "eu-west-1",
          },
        ],
        "SecretString": {
          "Ref": "DelegatedZoneCert71201BFD",
        },
      },
      "Type": "AWS::SecretsManager::Secret",
      "UpdateReplacePolicy": "Delete",
    },
    "CustomCrossAccountZoneDelegationCustomResourceProviderHandler44A84265": {
      "DependsOn": [
        "CustomCrossAccountZoneDelegationCustomResourceProviderRoleED64687B",
      ],
      "Properties": {
        "Code": {"S3Bucket":{"Fn::Sub":"cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}"},"S3Key":"[HASH REMOVED].zip"},
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomCrossAccountZoneDelegationCustomResourceProviderRoleED64687B",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomCrossAccountZoneDelegationCustomResourceProviderRoleED64687B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "DelegatedZoneCert71201BFD": {
      "Properties": {
        "DomainName": "*.test.awscommunitybuilders.org",
        "DomainValidationOptions": [
          {
            "DomainName": "*.test.awscommunitybuilders.org",
            "HostedZoneId": {
              "Ref": "delegatedZone93B2A299",
            },
          },
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "test-stack/DelegatedZoneCert",
          },
        ],
        "ValidationMethod": "DNS",
      },
      "Type": "AWS::CertificateManager::Certificate",
    },
    "DelegationRecordCrossAccountZoneDelegationCustomResource3FB1AD46": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "DelegationRecordcrossaccountzonedelegationhandlerrolePolicy7B31DBF8",
      ],
      "Properties": {
        "AssumeRoleArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":iam::9999:role/HostedZoneDelegationRole",
            ],
          ],
        },
        "DelegatedZoneName": "test.awscommunitybuilders.org",
        "DelegatedZoneNameServers": {
          "Fn::GetAtt": [
            "delegatedZone93B2A299",
            "NameServers",
          ],
        },
        "ParentZoneName": "awscommunitybuilders.org",
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomCrossAccountZoneDelegationCustomResourceProviderHandler44A84265",
            "Arn",
          ],
        },
        "TTL": 172800,
      },
      "Type": "Custom::CrossAccountZoneDelegation",
      "UpdateReplacePolicy": "Delete",
    },
    "DelegationRecordcrossaccountzonedelegationhandlerrolePolicy7B31DBF8": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":iam::9999:role/HostedZoneDelegationRole",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "DelegationRecordcrossaccountzonedelegationhandlerrolePolicy7B31DBF8",
        "Roles": [
          {
            "Fn::Select": [
              1,
              {
                "Fn::Split": [
                  "/",
                  {
                    "Fn::Select": [
                      5,
                      {
                        "Fn::Split": [
                          ":",
                          {
                            "Fn::GetAtt": [
                              "CustomCrossAccountZoneDelegationCustomResourceProviderRoleED64687B",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "delegatedZone93B2A299": {
      "Properties": {
        "Name": "test.awscommunitybuilders.org.",
      },
      "Type": "AWS::Route53::HostedZone",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
